data-plane/bindings/adapter/src/adapter.rs

```diff
@@ -0,0 +1,1841 @@
+            ServiceError::ReceiveError(msg) => SlimError::ReceiveError { message: msg },
+            ServiceError::SessionError(msg) => SlimError::SessionError { message: msg },
+            _ => SlimError::InternalError {
+                message: err.to_string(),
+            },
+        }
+    }
+}
+
+impl From<SlimSessionError> for SlimError {
+    fn from(err: SlimSessionError) -> Self {
+        SlimError::SessionError {
+            message: err.to_string(),
+        }
+    }
+}
+
+/// TLS configuration for server/client
+#[derive(uniffi::Record)]
+pub struct TlsConfig {
+    pub insecure: bool,
+}
+
+/// Server configuration for running a SLIM server
+#[derive(uniffi::Record)]
+pub struct ServerConfig {
+    pub endpoint: String,
+    pub tls: TlsConfig,
+}
+
+/// Client configuration for connecting to a SLIM server
+#[derive(uniffi::Record)]
+pub struct ClientConfig {
+    pub endpoint: String,
+    pub tls: TlsConfig,
+}
+
+/// Message context for received messages
+#[derive(uniffi::Record)]
+pub struct MessageContext {
+    pub source_name: Name,
+    pub destination_name: Option<Name>,
+    pub payload_type: String,
+    pub metadata: std::collections::HashMap<String, String>,
+    pub input_connection: u64,
+    pub identity: String,
+}
+
+impl From<crate::MessageContext> for MessageContext {
+    fn from(ctx: crate::MessageContext) -> Self {
+        Self {
+            source_name: Name::from(&ctx.source_name),
+            destination_name: ctx.destination_name.as_ref().map(Name::from),
+            payload_type: ctx.payload_type,
+            metadata: ctx.metadata,
+            input_connection: ctx.input_connection,
+            identity: ctx.identity,
+        }
+    }
+}
+
+impl From<MessageContext> for crate::MessageContext {
+    fn from(ctx: MessageContext) -> Self {
+        Self {
+            source_name: ctx.source_name.into(),
+            destination_name: ctx.destination_name.map(|n| n.into()),
+            payload_type: ctx.payload_type,
+            metadata: ctx.metadata,
+            input_connection: ctx.input_connection,
+            identity: ctx.identity,
+        }
+    }
+}
+
+/// Received message containing context and payload
+#[derive(uniffi::Record)]
+pub struct ReceivedMessage {
+    pub context: MessageContext,
+    pub payload: Vec<u8>,
+}
+
+// ============================================================================
+// FFI Entry Points
+// ============================================================================
+
+/// Create an app with the given name and shared secret (blocking version for FFI)
+///
+/// This is the main entry point for creating a SLIM application from language bindings.
+#[uniffi::export]
+pub fn create_app_with_secret(
+    app_name: Name,
+    shared_secret: String,
+) -> Result<Arc<BindingsAdapter>, SlimError> {
+    let runtime = get_runtime();
+    runtime.block_on(async { create_app_with_secret_async(app_name, shared_secret).await })
```

Comment:
The `create_app_with_secret_async` should not be async I think:

```suggestion
    create_app_with_secret_async(app_name, shared_secret)
```